### PR TITLE
Default `diff_type` to forward difference & fix Jacobian/W reuse logic

### PR DESCRIPTION
**Pull request by YingboMa**

### CODE REVIEW

**Comment by YingboMa** [comment]
PR:
```julia
julia> sol_cvodebdf = @time solve(prob, CVODE_BDF());
 29.649620 seconds (17.69 k allocations: 25.145 MiB, 0.02% gc time)

julia> sol_cvodebdf.destats
DiffEqBase.DEStats
Number of function 1 evaluations: 1521
Number of function 2 evaluations: -1
Number of W matrix evaluations: 169
Number of linear solves:        -1
Number of Jacobians created:        26
Number of nonlinear solver iterations:        1517
Number of nonlinear solver convergence failures:        0
Number of rootfind condition calls:        0
Number of accepted steps:       879
Number of rejected steps:       63

julia> sol_kencarp4 = @time solve(prob, KenCarp4(autodiff=false));
 41.321142 seconds (522.05 k allocations: 4.407 GiB, 1.63% gc time)

julia> sol_kencarp4.destats
DiffEqBase.DEStats
Number of function 1 evaluations: 518194
Number of function 2 evaluations: 0
Number of W matrix evaluations: 232
Number of linear solves:        4553
Number of Jacobians created:        229
Number of nonlinear solver iterations:        4315
Number of nonlinear solver convergence failures:        67
Number of rootfind condition calls:        0
Number of accepted steps:       238
Number of rejected steps:       0
```

**Comment by YingboMa** [comment]
```PR
julia> sol_kencarp4 = @time solve(prob, KenCarp4(autodiff=false));
 36.643173 seconds (3.30 k allocations: 27.780 MiB, 0.01% gc time)

julia> sol_kencarp4.destats
DiffEqBase.DEStats
Number of function 1 evaluations: 518194
Number of function 2 evaluations: 0
Number of W matrix evaluations: 232
Number of linear solves:        4553
Number of Jacobians created:        229
Number of nonlinear solver iterations:        4315
Number of nonlinear solver convergence failures:        67
Number of rootfind condition calls:        0
Number of accepted steps:       238
Number of rejected steps:       0
```

**Comment by YingboMa** [comment]
```julia
using DiffEqBiological, ReactionNetworkImporter
using OrdinaryDiffEq, Sundials

function getfun1()
  networkname = "testbcrbng"
  datadir     = joinpath(dirname(pathof(ReactionNetworkImporter)), "..", "data", "bcr")
  fname       = joinpath(datadir, "bcr.net")
  prnbng      = loadrxnetwork(BNGNetwork(),string(networkname,"bng"), fname);
  rn = prnbng.rn; u0 = prnbng.u₀; p = prnbng.p
  addodes!(rn; build_jac=false, build_symfuncs=false)
  rn, u0, p
end
rn, u0, p = getfun1()

du = similar(u0);
prob = ODEProblem(rn.f, u0, (0, 10000.), p)

@time prob.f(du, u0, p, 0.1)

sol_kencarp4 = @time solve(prob, KenCarp4(autodiff=false));
sol_kencarp4.destats
sol_cvodebdf = @time solve(prob, CVODE_BDF());
sol_cvodebdf.destats
```
```julia
julia> sol_kencarp4 = @time solve(prob, KenCarp4(autodiff=false, diff_type=Val{:forward}));
 25.876346 seconds (551.00 k allocations: 52.364 MiB, 0.04% gc time)

julia> sol_kencarp4.destats
DiffEqBase.DEStats
Number of function 1 evaluations: 261256
Number of function 2 evaluations: 0
Number of W matrix evaluations: 232
Number of linear solves:        4553
Number of Jacobians created:        229
Number of nonlinear solver iterations:        4315
Number of nonlinear solver convergence failures:        67
Number of rootfind condition calls:        0
Number of accepted steps:       238
Number of rejected steps:       0

julia> sol_cvodebdf = @time solve(prob, CVODE_BDF());
 27.196137 seconds (1.58 M allocations: 107.659 MiB, 0.72% gc time)

julia> sol_cvodebdf.destats
DiffEqBase.DEStats
Number of function 1 evaluations: 1521
Number of function 2 evaluations: -1
Number of W matrix evaluations: 169
Number of linear solves:        -1
Number of Jacobians created:        26
Number of nonlinear solver iterations:        1517
Number of nonlinear solver convergence failures:        0
Number of rootfind condition calls:        0
Number of accepted steps:       879
Number of rejected steps:       63
```

We are faster than CVODE_BDF now. @isaacsas

**Comment by YingboMa** [comment]
issue664

File: src/algorithms.jl [hunk 1]
```diff
@@ -736,101 +738,28 @@ struct ESDIRK54I8L2SA{CS,AD,F,F2,FDT,T2} <: OrdinaryDiffEqNewtonAdaptiveAlgorith
...
-struct Rodas5{CS,AD,F,FDT} <: OrdinaryDiffEqRosenbrockAdaptiveAlgorithm{CS,AD}
-  linsolve::F
-  diff_type::FDT
+for Alg in [:Rosenbrock23, :Rosenbrock32, :ROS3P, :Rodas3, :RosShamp4, :Veldd4, :Velds4, :GRK4T, :GRK4A, :Ros4LStab, :Rodas4, :Rodas42, :Rodas4P, :Rodas5]
+  @eval begin
+    struct $Alg{CS,AD,F,FDT} <: OrdinaryDiffEqRosenbrockAdaptiveAlgorithm{CS,AD}
+      linsolve::F
+      diff_type::FDT
+    end
+    $Alg(;chunk_size=0,autodiff=true,diff_type=Val{:central},linsolve=DEFAULT_LINSOLVE) = $Alg{chunk_size,autodiff,typeof(linsolve),typeof(diff_type)}(linsolve,diff_type)
```
Comment (isaacsas) [inline src/algorithms.jl]: Just curious, but why keep `:central` here?

Comment (YingboMa) [reply src/algorithms.jl]: Rosenbrock methods rely on accurate Jacobian.

Comment (isaacsas) [reply src/algorithms.jl]: Got it.

**Comment by isaacsas** [comment]
What are your SDIRK and BDF plots showing? Select solution components?

**Comment by YingboMa** [comment]
Yes, basically `sol.t` vs `sol.u`.

File: src/derivative_utils.jl [hunk 2]
```diff
@@ -237,82 +237,89 @@ function LinearAlgebra.mul!(Y::AbstractVecOrMat, W::WOperator, B::AbstractVecOrM
...
+    @label J2W
+    # skip calculation of W if step is repeated
+    ( new_W = donewW(integrator, new_jac) ) && (W.transform = W_transform; set_gamma!(W, dtgamma))
+  else # concrete W using jacobian from `calc_J!`
+    # skip calculation of J if step is repeated
+    ( new_jac = donewJ(integrator, alg, repeat_step) ) && calc_J!(integrator, cache, is_compos)
+    # skip calculation of W if step is repeated
+    ( new_W = donewW(integrator, new_jac) ) && jacobian2W!(W, mass_matrix, dtgamma, J, W_transform)
+  end
+  isnewton && set_new_W!(cache.nlsolver, new_W)
```
Comment (ChrisRackauckas) [inline src/derivative_utils.jl]: why is this needed?

File: src/algorithms.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/algorithms.jl]: What did our benchmarks end up showing here? I think :forward seemed to be enough? But what problem was it tested on?

**Comment by YingboMa** [comment]
rober

hires


For Rosenbrock methods, the central difference is better than the forward difference when a problem is very stiff.
